### PR TITLE
Homework 1

### PR DESCRIPTION
**Pull request by starostinak**

### CODE REVIEW

File: course5/homework1/homework1_hell.cc [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+#include <functional>
+using namespace std;
+
+const int MAX_NUM_ARGS = 445;
+
+template <int I>
+void print(void)
+{
+    print<I - 1>();
+    std::cout << I << " " << std::endl;
```
Comment (vargad) [inline course5/homework1/homework1_hell.cc]: Space is not needed here. Very fancy code, nightmare to godbolt :)

File: course5/homework1/homework1_v2.cc [hunk 2]
```diff
@@ -0,0 +1,13 @@
+#include <iostream>
+#include <algorithm>
+#include <cstring>
+
+int main(int argc, char** argv)
+{
+    std::sort(argv, argv + argc,
+            [](const char* s1, const char *s2){ return strcmp(s1, s2) < 0; });
```
Comment (maciekgajewski) [inline course5/homework1/homework1_v2.cc]: Use ```std::strcmp```. The unqualified version is a C-call, which is accessible only by accident.

File: course5/homework1/homework1_v2.cc [hunk 3]
```diff
@@ -0,0 +1,13 @@
+#include <iostream>
+#include <algorithm>
+#include <cstring>
+
+int main(int argc, char** argv)
+{
+    std::sort(argv, argv + argc,
```
Comment (maciekgajewski) [inline course5/homework1/homework1_v2.cc]: Have you looked at the output from the program?

File: course5/homework1/homework1_hell.cc [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+template <>
+void print<0>(void)
+{
+    std::cout << "0" << std::endl;
+}
+
+std::function<void(void)> printers[MAX_NUM_ARGS];
```
Comment (maciekgajewski) [inline course5/homework1/homework1_hell.cc]: You could've dress it in a class, instead of using global variable.

Comment (maciekgajewski) [inline course5/homework1/homework1_hell.cc]: Raw function pointer could shave few bytes from the data structure

File: course5/homework1/homework1_hell.cc [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+template<>
+void definePrinters<0>()
+{
+    printers[0] = &print<0>;
+}
+
+int main(int argc, char **argv)
+{
+    definePrinters<MAX_NUM_ARGS>();
+    printers[argc - 1]();
```
Comment (maciekgajewski) [inline course5/homework1/homework1_hell.cc]: You should fall-back to runtime iteration for argcs > MAX_NUM_ARGS

File: course5/homework1/homework1_hell.cc [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    return 0;
+}
```
Comment (maciekgajewski) [inline course5/homework1/homework1_hell.cc]: Total overkill, but with a potential. To justify it, you should provide a benchmark that proves it's faster than a for loop for some values of argc.

**Review by maciekgajewski — COMMENTED** [review]
Interesting, but needs work

**Pull request closed without merging**